lib/executor/ramping_arrival_rate.go

```diff
@@ -316,25 +316,52 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 	returnedVUs := make(chan struct{})
 	startTime, maxDurationCtx, regDurationCtx, cancel := getDurationContexts(parentCtx, duration, gracefulStop)
 
+	// Iterations chan notifies a new available iteration request.
+	// The number of active VU iterators and
+	// the rate computed should be able to maintain this chan empty most of the time.
+	// If the chan is not empty then
+	// it means the sender rate is not matching the VUs availability.
+	iterations := make(chan struct{}, 1)
+	defer close(iterations)
+
+	// exitVUIterator notifies when an iterator must stop to process.
+	// So, a message on the chan decreases the number of active iterators.
+	exitVUIterator := make(chan struct{}, maxVUs)
+
 	defer func() {
 		// Make sure all VUs aren't executing iterations anymore, for the cancel()
 		// below to deactivate them.
 		<-returnedVUs
 		cancel()
 		activeVUsWg.Wait()
+		close(exitVUIterator)
 	}()
+
 	activeVUs := make(chan lib.ActiveVU, maxVUs)
 	activeVUsCount := uint64(0)
 
 	returnVU := func(u lib.InitializedVU) {
+		exitVUIterator <- struct{}{}
 		varr.executionState.ReturnVU(u, true)
 		activeVUsWg.Done()
 	}
+
+	runIterationBasic := getIterationRunner(varr.executionState, varr.logger)
 	activateVU := func(initVU lib.InitializedVU) lib.ActiveVU {
 		activeVUsWg.Add(1)
 		activeVU := initVU.Activate(getVUActivationParams(maxDurationCtx, varr.config.BaseConfig, returnVU))
 		varr.executionState.ModCurrentlyActiveVUsCount(+1)
 		atomic.AddUint64(&activeVUsCount, 1)
+		go func() {
+			for {
+				select {
+				case <-iterations:
+					runIterationBasic(maxDurationCtx, activeVU)
+				case <-exitVUIterator:
```

Comment:
I am not sure if you even need a new `exitVUIterator` channel. As far as I understand your code, you are using this as a condition for VUs to stop listening for new iterations. If so, listening to `regDurationCtx.Done()` should be enough, right? Read [the docs for `getDurationContexts()`](https://github.com/k6io/k6/blob/v0.31.1/lib/executor/helpers.go#L116-L141) to get an idea what the contexts are.